File: tool/feedback/src/main/java/feedback/parser/ParserUtil.java
package feedback.parser;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Stream;

public final class ParserUtil {
    public static String[] getFileLines(final Path path) throws IOException {
        final List<String> lines = new ArrayList<>();
        try (final Stream<String> stream = Files.lines(path)) {
            stream.forEach(lines::add);
        }
        return lines.toArray(new String[0]);
    }

    public static String[] getFileLines(final String path) throws IOException {
        return getFileLines(Paths.get(path));
    }
}

File: tool/feedback/src/main/scala/feedback/parser/ParserUtil.scala
package feedback.parser

import feedback.{AsyncIterator, ThreadUtil}

import java.io.File
import java.nio.file.{Files, Path}
import scala.io.Source
import scala.util.{Success, Using}

object ParserUtil {
  private val MAX_FILE_SIZE = 200_000_000L  // in bytes

  def getFileLines(file: File): Array[String] = {
    require(Files.size(file.toPath) < MAX_FILE_SIZE)
    Using(Source.fromFile(file)) {_.getLines().toArray} match {
      case Success(lines) => lines
    }
  }

  def getFileLines(path: Path): Array[String] = getFileLines(path.toFile)

  def getFileLines(path: String): Array[String] = getFileLines(new File(path))

  def getFileLinesAsync(file: File): Iterator[String] = {
    require(Files.size(file.toPath) < MAX_FILE_SIZE)
    new AsyncIterator[String](add => Using(Source.fromFile(file)) { file =>
      val iterator = file.getLines()
      while (iterator.hasNext) {
        add(iterator.next())
      }
    } match {
      case Success(_) =>
    })
  }

  def getFileLinesAsync(path: Path): Iterator[String] = getFileLinesAsync(path.toFile)
}
